File: sources/net.sf.j2s.core/test/dev/j2s/org/ultrastudio/boids/Boid.js
Clazz.declarePackage ("org.ultrastudio.boids");
Clazz.load (["java.awt.geom.Point2D"], "org.ultrastudio.boids.Boid", ["org.ultrastudio.boids.Constants"], function () {
c$ = Clazz.decorateAsClass (function () {
this.location = null;
this.speed = null;
this.flock = null;
Clazz.instantialize (this, arguments);
}, org.ultrastudio.boids, "Boid");
Clazz.prepareFields (c$, function () {
this.location =  new java.awt.geom.Point2D.Double (100 * Math.random (), 100 * Math.random ());
this.speed =  new java.awt.geom.Point2D.Double (Math.random (), Math.random ());
});
Clazz.defineMethod (c$, "moveAhead", 
function (time) {
this.location.x += this.speed.x * time * org.ultrastudio.boids.Constants.stepTime;
this.location.y += this.speed.y * time * org.ultrastudio.boids.Constants.stepTime;
if (this.flock.cooperative) {
if (!this.flock.scared) this.matchSpeed (time);
this.keepSpeed (time);
this.flyTowardsCenterOfMass (time);
}this.keepAway (time);
this.bounceFromWalls ();
}, "~N");
Clazz.defineMethod (c$, "matchSpeed", 
function (time) {
var x = 0;
var y = 0;
var n = this.flock.size () - 1;
for (var boid, $boid = this.flock.iterator (); $boid.hasNext () && ((boid = $boid.next ()) || true);) if (boid !== this) {
x += boid.speed.x;
y += boid.speed.y;
}
x /= n;
y /= n;
var vx = x - this.speed.x;
var vy = y - this.speed.y;
this.speed.x += vx * org.ultrastudio.boids.Constants.speedAdjust * time;
this.speed.y += vy * org.ultrastudio.boids.Constants.speedAdjust * time;
}, "~N");
Clazz.defineMethod (c$, "flyTowardsCenterOfMass", 
function (time) {
var x = 0;
var y = 0;
var n = this.flock.size () - 1;
for (var boid, $boid = this.flock.iterator (); $boid.hasNext () && ((boid = $boid.next ()) || true);) if (boid !== this) {
x += boid.location.x;
y += boid.location.y;
}
x /= n;
y /= n;
var vx = x - this.location.x;
var vy = y - this.location.y;
if (this.flock.scared) {
vx = -vx;
vy = -vy;
}this.move (time, vx, vy, org.ultrastudio.boids.Constants.holdFlockCenter);
}, "~N");
Clazz.defineMethod (c$, "keepSpeed", 
function (time) {
var v = Math.sqrt (this.speed.x * this.speed.x + this.speed.y * this.speed.y);
var r = org.ultrastudio.boids.Constants.minSpeed - v;
if (r > 0 && v < org.ultrastudio.boids.Constants.maxSpeed) r = 1 + org.ultrastudio.boids.Constants.speedAdjust;
 else r = 1 - org.ultrastudio.boids.Constants.speedAdjust;
this.speed.x *= r;
this.speed.y *= r;
}, "~N");
Clazz.defineMethod (c$, "move", 
 function (time, vx, vy, weight) {
this.location.x += vx * weight * time;
this.location.y += vy * weight * time;
}, "~N,~N,~N,~N");
Clazz.defineMethod (c$, "keepAway", 
function (time) {
var vx = 0;
var vy = 0;
for (var boid, $boid = this.flock.iterator (); $boid.hasNext () && ((boid = $boid.next ()) || true);) {
var distance = boid.location.distance (this.location);
if (boid !== this && distance < org.ultrastudio.boids.Constants.keepAwayDistance) {
if (distance < 1) distance = 1;
var s = 1 / distance;
vx += s * (this.location.x - boid.location.x);
vy += s * (this.location.y - boid.location.y);
}}
this.move (time, vx, vy, org.ultrastudio.boids.Constants.avoidNeibourthood);
}, "~N");
Clazz.defineMethod (c$, "bounceFromWalls", 
function () {
if (this.location.x < org.ultrastudio.boids.Constants.keepAwayDistance) {
this.speed.x += org.ultrastudio.boids.Constants.avoidNeibourthood;
}if (this.location.y < org.ultrastudio.boids.Constants.keepAwayDistance) {
this.speed.y += org.ultrastudio.boids.Constants.avoidNeibourthood;
}if (this.location.x > this.flock.$$size.width - org.ultrastudio.boids.Constants.keepAwayDistance) {
this.speed.x -= org.ultrastudio.boids.Constants.avoidNeibourthood;
}if (this.location.y > this.flock.$$size.height - org.ultrastudio.boids.Constants.keepAwayDistance) {
this.speed.y -= org.ultrastudio.boids.Constants.avoidNeibourthood;
}if (this.location.x < 0) {
this.location.x = 0;
this.speed.x = -this.speed.x;
}if (this.location.y < 0) {
this.location.y = 0;
this.speed.y = -this.speed.y;
}if (this.location.x > this.flock.$$size.width) {
this.location.x = this.flock.$$size.width;
this.speed.x = -this.speed.x;
}if (this.location.y > this.flock.$$size.height) {
this.location.y = this.flock.$$size.height;
this.speed.y = -this.speed.y;
}});
});

File: sources/net.sf.j2s.java.core/src/org/ultrastudio/boids/Boid.java
/*
 Copyright (C) 2010  Audrius Meskauskas
 Flock behavior simulator

 This  program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published  by
 the Free Software Foundation; either version 2 of the License, or (at
 your option) any later version.

 This program is distributed in the hope that it will be  useful,  but
 WITHOUT   ANY   WARRANTY;   without  even  the  implied  warranty  of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR  PURPOSE.   See  the  GNU
 General Public License for more details.

 You  should  have  received  a copy of the GNU General Public License
 along  with  this  program;  if  not,  write  to  the  Free  Software
 Foundation,  Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 USA.
*/
 
package org.ultrastudio.boids;

import java.awt.geom.Point2D;

/**
 * A single boid
 *
 *  More on Boids can be found at 
 *  
 *  http://ultrastudio.org/en/Flock
 *  http://www.vergenet.net/~conrad/boids/
 * 
 * @author Audrius Meskauskas
 */
public class Boid {

	/**
	 * The boid location (fields are modified the the object is never reallocated hence final)
	 */
	final Point2D.Double location = new Point2D.Double(100 * Math.random(),
			100 * Math.random());

	/**
	 * The boid flight vector
	 */
	final Point2D.Double speed = new Point2D.Double(Math.random(), Math
			.random());

	/**
	 * The flock where this boid belongs. The flock is not final, can theoretically change.
	 */
	Flock flock;

	/**
	 * Fly in the direction, determined by the flight vector.
	 * 
	 * @param time the duration of the model iteration. 
	 */
	public void moveAhead(double time) {
		location.x += speed.x * time * Constants.stepTime;
		location.y += speed.y * time * Constants.stepTime;

		if (flock.cooperative) {
			if (!flock.scared)
				matchSpeed(time);
			keepSpeed(time);
			flyTowardsCenterOfMass(time);
		}
		keepAway(time);
		bounceFromWalls();
	}

	/**
	 * Try to keep the speed the same as speed of other boids.
	 */
	protected void matchSpeed(double time) {
		// Compute the perceive center.
		double x = 0;
		double y = 0;

		// Do not count self
		int n = flock.size() - 1;
		for (Boid boid : flock)
			if (boid != this) {
				x += boid.speed.x;
				y += boid.speed.y;
			}

		x /= n;
		y /= n;

		// Compute the vector toward center
		double vx = x - speed.x;
		double vy = y - speed.y;

		speed.x += vx * Constants.speedAdjust * time;
		speed.y += vy * Constants.speedAdjust * time;
	}

	/**
	 * Fly towards the perceived center of mass
	 */
	protected void flyTowardsCenterOfMass(double time) {

		// Compute the perceive center.
		double x = 0;
		double y = 0;

		// Do not count self
		int n = flock.size() - 1;
		for (Boid boid : flock)
			if (boid != this) {
				x += boid.location.x;
				y += boid.location.y;
			}

		x /= n;
		y /= n;

		// Compute the vector toward center
		double vx = x - location.x;
		double vy = y - location.y;

		// If scared, fly AWAY from the center of mass
		if (flock.scared) {
			vx = -vx;
			vy = -vy;
		}

		move(time, vx, vy, Constants.holdFlockCenter);
	}

	/**
	 * Accelerate if too slow.
	 */
	public void keepSpeed(double time) {

		double v = Math.sqrt(speed.x * speed.x + speed.y * speed.y);
		double r = Constants.minSpeed - v;

		if (r > 0 && v < Constants.maxSpeed)
			r = 1 + Constants.speedAdjust;
		else
			r = 1 - Constants.speedAdjust;

		speed.x *= r;
		speed.y *= r;
	}

	/**
	 * Service method to move
	 * @param time the time quant
	 * @param vx x speed vector
	 * @param vy y speed vector
	 * @param weight the weight of this move
	 */
	private void move(double time, double vx, double vy, double weight) {
		location.x += vx * weight * time;
		location.y += vy * weight * time;
	}

	/**
	 * Keep away from walls and other boids
	 */
	protected void keepAway(double time) {
		double vx = 0;
		double vy = 0;

		// Do not try to keep away from himself
		for (Boid boid : flock) {
			double distance = boid.location.distance(location);
			if (boid != this && distance < Constants.keepAwayDistance) {
				// Handle degenerate case when boids are at the same location.
				if (distance < 1)
					distance = 1;

				// Make force inverse proportional to the distance.
				double s = 1 / distance;
				vx += s * (location.x - boid.location.x);
				vy += s * (location.y - boid.location.y);
			}
		}
		move(time, vx, vy, Constants.avoidNeibourthood);
	}

	/**
	 * Bounce from the boundaries of the modeling space (like bats in the cave). To avoid sticking on the boundary
	 * (altering vector on every iteration), this must be the last method to call.
	 */
	protected void bounceFromWalls() {

		// Soft bounce (keep away)
		if (location.x < Constants.keepAwayDistance) {
			speed.x += Constants.avoidNeibourthood;
		}

		if (location.y < Constants.keepAwayDistance) {
			speed.y += Constants.avoidNeibourthood;
		}

		if (location.x > flock.size.width - Constants.keepAwayDistance) {
			speed.x -= Constants.avoidNeibourthood;
		}

		if (location.y > flock.size.height - Constants.keepAwayDistance) {
			speed.y -= Constants.avoidNeibourthood;
		}

		// Hard bounce
		if (location.x < 0) {
			location.x = 0;
			speed.x = -speed.x;
		}

		if (location.y < 0) {
			location.y = 0;
			speed.y = -speed.y;
		}

		if (location.x > flock.size.width) {
			location.x = flock.size.width;
			speed.x = -speed.x;
		}

		if (location.y > flock.size.height) {
			location.y = flock.size.height;
			speed.y = -speed.y;
		}
	}

}
